src/diffTool/rulesProcessor.ts

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright (c) 2020, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+import {
+  Engine,
+  RuleProperties,
+  Event,
+  Almanac,
+  RuleResult,
+  EngineResult
+} from "json-rules-engine";
+import { NodeDiff } from "./jsonDiff";
+import fs from "fs-extra";
+import { ramlToolLogger } from "../logger";
+/* eslint-disable @typescript-eslint/no-use-before-define */
+
+//ID for the diff node/fact that is passed to the rule engine
+export const DIFF_FACT_ID = "diff";
+
+/**
+ * Apply rules on the AMF node differences, updates passed rules details in the node
+ * @param diffs - Array of differences
+ * @param rulesPath - File path of the rules
+ */
+export async function applyRules(
+  diffs: NodeDiff[],
+  rulesPath: string
+): Promise<void> {
+  if (!diffs || diffs.length == 0) {
+    ramlToolLogger.info("No differences to apply the rules");
+    return;
+  }
+  const rules = getRules(rulesPath);
+  if (!rules || rules.length == 0) {
+    ramlToolLogger.info("No rules to apply on the differences");
+    return;
+  }
+  //initialize engine with rules
+  const engine = new Engine(rules);
+  //callback function to execute when a rule is passed/evaluates to true
+  engine.on("success", successHandler);
+  addCustomOperators(engine);
+
+  //run rules on diffs
+  const promises = diffs.map(diff => {
+    return runEngine(engine, diff);
+  });
+  await Promise.all(promises);
+}
+
+/**
+ * Get rules from rules json file
+ * @param rulesPath - Path to rules json file
+ *
+ * @returns Array of rules
+ */
+function getRules(rulesPath: string): RuleProperties[] {
+  if (!rulesPath) {
+    throw new Error(`Invalid rules path: ${rulesPath}`);
+  }
+  let rules: RuleProperties[];
+  try {
+    rules = fs.readJSONSync(rulesPath);
+  } catch (error) {
+    error.message = `Error parsing the rules file '${rulesPath}': ${error.message}`;
+    throw error;
+  }
+  if (!Array.isArray(rules)) {
+    throw new Error("Rules must be defined as a json array");
+  }
+  return rules;
+}
+
+/**
+ * Callback function that executes when a rule is passed/evaluates to true on a diff
+ * @param event - Event defined in the rule
+ * @param almanac - Almanac that has the fact/diff on which the rule is applied
+ * @param ruleResult - Result of rule execution
+ */
+async function successHandler(
+  event: Event,
+  almanac: Almanac,
+  ruleResult: RuleResult
+): Promise<void> {
+  const nodeDiff = (await almanac.factValue(DIFF_FACT_ID)) as NodeDiff;
+  ramlToolLogger.debug(
+    `Rule '${ruleResult.name}' is passed on difference '${nodeDiff.id}'`
+  );
+  nodeDiff.rule = {
```

Comment:
What is this assignment doing? It is at the end of the method and I don't see how it is being used.

Reply:
Adds rule that is passed/evaluated to true to the diff object